plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,955 @@
+
+    """
+
+    # Catch an error that can occur if a mode requires a gradient but a
+    # non-uniform grid is supplied, since this is not supported currently.
+    uses_grad = ["electrostatic gaussian sphere", "electrostatic spherical shock"]
+    if regular_grid is False and mode in uses_grad:
+        raise ValueError(
+            f"The mode {mode} is not compatible with a ",
+            "non-uniform grid: please chose a different mode.",
+        )
+
+    # Process different options for inputs
+    if isinstance(num, int):
+        num = (num, num, num)
+
+    L = np.zeros([2, 3]) * length.unit
+    if length.size == 1:
+        L[:, 0] = np.array([-length.value, length.value]) * length.unit
+        L[:, 1] = np.array([-length.value, length.value]) * length.unit
+        L[:, 2] = np.array([-length.value, length.value]) * length.unit
+    elif length.size == 3:
+        L[:, 0] = np.array([-length[0].value, length[0].value]) * length.unit
+        L[:, 1] = np.array([-length[1].value, length[1].value]) * length.unit
+        L[:, 2] = np.array([-length[2].value, length[2].value]) * length.unit
+    else:
+        raise ValueError("Dimensions of length in test_fields are" " not valid.")
+
+    # If no grid is specified, create a grid
+    if grid is None:
+        grid = np.zeros([num[0], num[1], num[2], 3]) * u.cm
+
+        # If regulr_grid keyword is False, create a non-uniform grid with
+        # random point spacing in all directions
+        if not regular_grid:
+            for d in [0, 1, 2]:
+                ax = np.random.uniform(low=L[0, d].value, high=L[1, d].value, size=num)
+                ax = np.sort(ax, axis=d)
+                grid[..., d] = ax * L.unit
+
+        # If regular_grid keyword is set, create a uniformly spaced grid
+        else:
+            xaxis = np.linspace(L[0, 0], L[1, 0], num=num[0])
+            yaxis = np.linspace(L[0, 1], L[1, 1], num=num[1])
+            zaxis = np.linspace(L[0, 2], L[1, 2], num=num[2])
+            xarr, yarr, zarr = np.meshgrid(xaxis, yaxis, zaxis, indexing="ij")
+            grid[..., 0] = xarr
+            grid[..., 1] = yarr
+            grid[..., 2] = zarr
+
+    # Create axes and arrays of each coordinate from the grid
+    xaxis = grid[:, 0, 0, 0]
+    yaxis = grid[0, :, 0, 1]
+    zaxis = grid[0, 0, :, 2]
+    xarr = grid[..., 0]
+    yarr = grid[..., 1]
+    zarr = grid[..., 2]
+
+    # Calculate length of each axis, for use in scaling field features
+    length = np.abs(L[0, :] - L[1, :])
+
+    # Calculate radius arrays in spherical and cylindrical coordinates
+    # for use in generating different field structures
+    radius = np.sqrt(xarr ** 2 + yarr ** 2 + zarr ** 2)
+    pradius = np.sqrt(xarr ** 2 + yarr ** 2)
+
+    # Create blank arrays for fields
+    E = np.zeros(grid.shape) * u.V / u.m
+    B = np.zeros(grid.shape) * u.T
+
+    if mode == "no fields":
+        pass
+
+    elif mode == "electrostatic gaussian sphere":
```

Comment:
I was reading really recently about a refactoring strategy on [replacing conditional statements with polymorphism](https://www.codementor.io/@uditrastogi/replace-conditional-statements-if-else-or-switch-with-polymorphism-ryl8mx4ns).  One difficulty with `if`/`elif`/`else` blocks is that, if there end up being a lot of options, they can lead to code that is harder to maintain (as I've been repeatedly discovering for another code I've been using with conditional blocks that are hundreds of lines long...and that I've added hundreds of lines to...).  The link above goes through an interesting example of how object-oriented programming allows us to avoid that with [polymorphism](https://www.educative.io/edpresso/what-is-polymorphism)...in particular, creating different classes to represent the different situations.  I'm wondering if that would be a worthwhile approach here.  Hm...

Reply:
This sounds like it could work: I'll play around with it.

Reply:
If you want a python example, I basically did this with the fit functions for my `find_floating_potential()` in PR #889. The floating potential function allows you to choose two different curves to fit to the data.  There's a lot of common functionality needed between the two fits (fitted parameters, rsq, the func, a solver, etc.) but not all functionality is performed the same way.  For example, the solver and root finder for each curve use different methods. Thus, I package the fit functions into a set of parent, child classes that share a common API.  You can find the `FitFunctions` classes [here](https://github.com/rocco8773/PlasmaPy/blob/ad-lang-vf/plasmapy/analysis/swept_langmuir/fit_functions.py).  The "conditional selection" of the functions inside `find_floating_potential()` are [here](https://github.com/rocco8773/PlasmaPy/blob/e17b15aaa8a2f57e0f8c891c1caf2ded87415d9c/plasmapy/analysis/swept_langmuir/floating_potential.py#L146-L164).  And, the execution of the `FitFunctions` inside `find_floating_potential()` is [here](https://github.com/rocco8773/PlasmaPy/blob/e17b15aaa8a2f57e0f8c891c1caf2ded87415d9c/plasmapy/analysis/swept_langmuir/floating_potential.py#L296-L304).

Reply:
@namurphy @rocco8773  Thanks for the suggestion and example: I've implemented something like this in the code so let me know what you think of the new structure.

Reply:
I'll take a look at it.  I'm aiming to start a review later today.